sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerConfig.java

```diff
@@ -136,13 +163,51 @@ public SpannerConfig withHost(ValueProvider<String> host) {
     return toBuilder().setHost(host).build();
   }
 
+  public SpannerConfig withCommitDeadline(Duration commitDeadline) {
+    return withCommitDeadline(ValueProvider.StaticValueProvider.of(commitDeadline));
+  }
+
+  public SpannerConfig withCommitDeadline(ValueProvider<Duration> commitDeadline) {
+    return toBuilder().setCommitDeadline(commitDeadline).build();
+  }
+
+  public SpannerConfig withMaxCumulativeBackoff(Duration maxCumulativeBackoff) {
+    return withMaxCumulativeBackoff(ValueProvider.StaticValueProvider.of(maxCumulativeBackoff));
+  }
+
+  public SpannerConfig withMaxCumulativeBackoff(ValueProvider<Duration> maxCumulativeBackoff) {
+    return toBuilder().setMaxCumulativeBackoff(maxCumulativeBackoff).build();
+  }
+
   @VisibleForTesting
   SpannerConfig withServiceFactory(ServiceFactory<Spanner, SpannerOptions> serviceFactory) {
     return toBuilder().setServiceFactory(serviceFactory).build();
   }
 
   public SpannerAccessor connectToSpanner() {
     SpannerOptions.Builder builder = SpannerOptions.newBuilder();
+
+    if (getCommitDeadline() != null && getCommitDeadline().get().getMillis() > 0) {
+
+      // In Spanner API version 1.21 or above, we can set the deadline / total Timeout on an API
+      // call using the following code:
+      //
+      // UnaryCallSettings.Builder commitSettings =
+      // builder.getSpannerStubSettingsBuilder().commitSettings();
+      // RetrySettings.Builder commitRetrySettings = commitSettings.getRetrySettings().toBuilder()
+      // commitSettings.setRetrySettings(
+      //     commitRetrySettings.setTotalTimeout(
+      //         Duration.ofMillis(getCommitDeadlineMillis().get()))
+      //     .build());
+      //
+      // However, at time of this commit, the Spanner API is at only at v1.6.0, where the only
+      // method to set a deadline is with GRPC Interceptors, so we have to use that...
+      SpannerInterceptorProvider interceptorProvider =
+          SpannerInterceptorProvider.createDefault()
+              .with(new CommitDeadlineSettingInterceptor(getCommitDeadline().get()));
```

Comment:
Just to confirm, this deadline will not cause Dataflow workitems to fail but just that request will be retried by SpannerIO within the same workitem ? (otherwise jobs can significantly slow down or fail if we rely on workitem retries.)

Reply:
This seems to be the case. I'll merge after tests pass. Thanks folks.

Reply:
> Just to confirm, this deadline will not cause Dataflow workitems to fail but just that request will be retried by SpannerIO within the same workitem

Correct, it will backoff/retry up to a configurable time limit (default 15 mins per workitem).